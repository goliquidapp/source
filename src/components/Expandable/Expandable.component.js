import React, {Component} from 'react';
import {View, Text, TouchableOpacity, ScrollView} from 'react-native';
import {Icon} from 'react-native-elements';

import Colors from '../../resources/Colors.js';
import Themes from '../../resources/Theme.js';

export default class Expandable extends Component{
	state={open:false}
	renderHead=()=>{
		const {containerStyle, bodyStyle, headerStyle, titleStyle, title, icon, rightComponent}=this.props;
		const {open}=this.state;
		return (
				<TouchableOpacity style={[styles.header,headerStyle]} onPress={this.toggle} onLongPress={this.props.onLongPress}>
					<Text style={[styles.title,titleStyle]}>{title}</Text>
					<View style={styles.headerRight}>
							{icon&&<Icon name={icon.name} type={icon.type} color={Colors['White']} size={20}/>}
							{rightComponent}
							{
								open?
								<Icon name="chevron-up" type="entypo" color={Colors['White']} size={20}/>:
								<Icon name="chevron-down" type="entypo" color={Colors['White']} size={20}/>
							}
					</View>
				</TouchableOpacity>
			)
	}
	renderBody=()=>{
		const {containerStyle,bodyStyle, body}=this.props;
		const {open}=this.state;
		const keys=Object.keys(body)
		if (open)
			return(
					<View style={[styles.body,bodyStyle]}>
						<ScrollView>
							{keys.map((key,index)=>{
								if (typeof(body[key])==='object'){
									return (<Text key={index.toString()} style={styles.row}><Text style={styles.rowKey}>{key}</Text>: {JSON.stringify(body[key])}</Text>)
								}
								else return (<Text key={index.toString()} style={styles.row}><Text style={styles.rowKey}>{key}</Text>: {body[key]}</Text>)
							})}
						</ScrollView>
					</View>
				)
		else 
			return <View></View>
	}
	toggle=()=>{
		this.setState({open:!this.state.open})
	}
	render(){
		const {containerStyle, bodyStyle, headerStyle, titleStyle, title, icon}=this.props;
		const {open}=this.state;
		return (
				<View style={[styles.container,containerStyle]}>
					{this.renderHead()}
					{this.renderBody()}
				</View>
			)
	}
}

const styles={
	container:{
		width:'100%',
		alignSelf:'center',
		paddingHorizontal:10,
		marginVertical:5
	},
	body:{
		elevation:1,
		padding:10,
		backgroundColor:Colors['Desert Storm']
	},
	header:{
		flexDirection:'row',
		alignItems:'center',
		justifyContent:'space-between',
		elevation:2,
		fontSize:12,
		width:'100%',
		padding:20,
		alignSelf:'center',
		backgroundColor:Themes['dark'].primary1
	},
	headerRight:{
		flexDirection:'row',
		alignItems:'center',
		justifyContent:'flex-end',
		width:'40%'
	},
	row:{
		width:'100%',
		fontSize:14,
		paddingVertical:3,
		borderBottomWidth:1,
		borderBottomColor:Colors['Mercury']
	},
	rowKey:{
		fontWeight:'bold'
	},
	title:{
		color:Colors['White'],
		fontSize:16,
		width:'60%'
	}
}