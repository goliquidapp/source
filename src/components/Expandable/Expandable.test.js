import React from 'react';
import { Text, View, TouchableOpacity, ScrollView} from 'react-native';
import renderer from 'react-test-renderer';

import { shallow } from 'enzyme';

import Expandable from './Expandable.component';

describe("Testing Expandable component",()=>{
	it('renders correctly', () => {
		const dummyData={
			key1:'value1',
			key2:'value2'
		}

		const wrapper = renderer.create(
    		<Expandable title={"Click me to see details"} body={dummyData}/>
    	).toJSON();

		expect(wrapper).toMatchSnapshot();
	});

	it('extends after click', () => {
		const onPressEvent = jest.fn();
        onPressEvent.mockReturnValue('Body is shown');

		const dummyData={
			key1:'value1',
			key2:'value2'
		}

		const wrapper = shallow(
    		<Expandable title={"Click me to see details"} body={dummyData}/>
    	);

    	expect(wrapper.find(ScrollView).length).toEqual(0);

    	wrapper.find(TouchableOpacity).first().props().onPress();
    	
		expect(wrapper.find(ScrollView).length).toEqual(1);
	});

	it('calls on long press', () => {
		const onPressEvent = jest.fn();
        onPressEvent.mockReturnValue('Long pressed, showing menu');

		const dummyData={
			key1:'value1',
			key2:'value2'
		}

		const wrapper = shallow(
    		<Expandable title={"Click me to see details"} body={dummyData} onLongPress={onPressEvent}/>
    	);

    	wrapper.find(TouchableOpacity).first().props().onLongPress();
    	
		expect(onPressEvent.mock.calls.length).toBe(1);
	});
})