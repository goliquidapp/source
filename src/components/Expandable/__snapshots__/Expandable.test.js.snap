// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Expandable component renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignSelf": "center",
        "marginVertical": 5,
        "paddingHorizontal": 10,
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <View
    accessible={true}
    clickable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#22222C",
        "elevation": 2,
        "flexDirection": "row",
        "fontSize": 12,
        "justifyContent": "space-between",
        "opacity": 1,
        "padding": 20,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 16,
            "width": "60%",
          },
          undefined,
        ]
      }
    >
      Click me to see details
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "flex-end",
          "width": "40%",
        }
      }
    >
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "justifyContent": "center",
            }
          }
          theme={
            Object {
              "colors": Object {
                "disabled": "hsl(208, 8%, 90%)",
                "divider": "#bcbbc1",
                "error": "#ff190c",
                "grey0": "#393e42",
                "grey1": "#43484d",
                "grey2": "#5e6977",
                "grey3": "#86939e",
                "grey4": "#bdc6cf",
                "grey5": "#e1e8ee",
                "greyOutline": "#bbb",
                "platform": Object {
                  "android": Object {
                    "error": "#f44336",
                    "primary": "#2196f3",
                    "secondary": "#9C27B0",
                    "success": "#4caf50",
                    "warning": "#ffeb3b",
                  },
                  "ios": Object {
                    "error": "#ff3b30",
                    "primary": "#007aff",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                },
                "primary": "#2089dc",
                "searchBg": "#303337",
                "secondary": "#8F0CE8",
                "success": "#52c41a",
                "warning": "#faad14",
              },
            }
          }
          underlayColor="white"
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#FFFFFF",
                  "fontSize": 20,
                },
                Object {
                  "backgroundColor": "transparent",
                },
                Object {
                  "fontFamily": "Entypo",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
            testID="iconIcon"
          >
            ÔÖÅ
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View />
</View>
`;
